engines/real-time-ensembler/pyfunc_ensembler_runner/ensembler_runner.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import pandas as pd
+
+from typing import Dict, List, Any
+from turing.ensembler import PyFunc
+from mlflow import pyfunc
+
+
+class PyFuncEnsemblerRunner:
+    """
+    PyFunc ensembler runner used for real-time outputs
+    """
+
+    def __init__(self, artifact_dir: str):
+        self.artifact_dir = artifact_dir
+        self._ensembler = None
+
+    def load(self):
+        self._ensembler = pyfunc.load_model(self.artifact_dir)
+
+    def predict(self, inputs: Dict[str, Any]) -> List[Any]:
+        logging.info(f"Input request payload: {inputs}")
+        ensembler_inputs = PyFuncEnsemblerRunner._preprocess_input(inputs)
+        output = self._ensembler.predict(ensembler_inputs).iloc[0].to_list()
+        logging.info(f"Output response: {output}")
+        return output
+
+    @staticmethod
+    def _preprocess_input(inputs: Dict[str, Any]) -> pd.DataFrame:
+        features = pd.Series(PyFuncEnsemblerRunner._get_features_from_inputs(inputs))
+        predictions = pd.Series(PyFuncEnsemblerRunner._get_predictions_from_inputs(inputs))
+        treatment_config = pd.Series(PyFuncEnsemblerRunner._get_treatment_config_from_inputs(inputs))
+        preprocessed_input = pd.concat([features, predictions, treatment_config]).to_frame().transpose()
+        return preprocessed_input
+
+    @staticmethod
+    def _get_features_from_inputs(inputs: Dict[str, Any]) -> Dict[str, Any]:
+        features = PyFuncEnsemblerRunner._flatten_json(inputs['request'])
+        return features
+
+    @staticmethod
+    def _get_predictions_from_inputs(inputs: Dict[str, Any]) -> Dict[str, Any]:
+        raw_predictions = PyFuncEnsemblerRunner._flatten_json(inputs['response']['route_responses'])
+        predictions = PyFuncEnsemblerRunner._create_dict_with_headers(raw_predictions,
+                                                                      PyFunc.PREDICTION_COLUMN_PREFIX)
+        return predictions
+
+    @staticmethod
+    def _get_treatment_config_from_inputs(inputs: Dict[str, Any]) -> Dict[str, Any]:
+        raw_predictions = PyFuncEnsemblerRunner._flatten_json(inputs['response']['experiment'])
+        treatment_config = PyFuncEnsemblerRunner._create_dict_with_headers(raw_predictions,
+                                                                           PyFunc.TREATMENT_CONFIG_COLUMN_PREFIX)
+        return treatment_config
+
+    @staticmethod
+    def _flatten_json(y: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Helper function to normalise a nested dictionary into a dictionary of depth 1 with names following the
+        convention: key_1.key_2.key_3..., with a period acting as a delimiter between nested keys
+
+        Items in lists have their names rendered using their index numbers within the lists they are found in.
+        """
```

Comment:
I'm not very comfortable with this. It seems like `dict` could be a better container for this data, compared to `pandas.DataFrame/pandas.Series`. The reason for this is:
1. Latency – for real-time ensemblers, the latency requirement is much more critical, compared to batch ensembling jobs. Having all this json manipulation and transformation into a pandas containers is likely something we'd better avoid
2. This "flat-json" is not something the end-users would expect to work with. Also, JSON keys could have `.` in their names, which is not handled here

With that said, re-working the interface of the PyFunc ensembler makes more sense, and then `batch-ensembler` can be updated too, to transform pandas data into `dict`

Reply:
Okay with regards to the latency concerns for real-time ensemblers, I've reworked (overloaded) the `predict` method of the `PyFunc` base class a little to make it work differently depending on whether it gets passed a `dict` (when called by a real-time ensembler engine) or a `pandas.DataFrame` (when called by a batch-ensembler engine) as its argument:

https://github.com/deadlycoconuts/turing/blob/1bce96a181324636367236df3e9ad6d98c285c68/sdk/turing/ensembler.py#L73

With this change, I've completely removed any of those `pandas.DataFrame` transformations like `flatten_json` or other preprocessing table operations that might introduce latency in the real-time ensemblers.

---
While it'd be nice to unify both the batch and real-time ensembling use cases to act on and return `dict` inputs, the `predict` method, as an inherited method of the `mlflow.pyfunc.PythonModel` abstract class, gets utilised by other downstream methods such as `mlflow.pyfunc.spark_udf`, that forces the `predict` method to follow a certain contract by taking `pandas.DataFrame`s as input and output. 

In particular, the current batch ensembling engine uses `mlflow.pyfunc.spark_udf`, which necessitates at least some implementation of `predict` that takes in and outputs `pandas.DataFrame`s. For now I've kept our original implementation intact to prevent the batch ensembler from breaking, but I'll see if I can find a solution that's more elegant.